src/ButteredBread.sol

```diff
@@ -123,25 +126,27 @@ contract ButteredBread is IButteredBread, ERC20VotesUpgradeable, OwnableUpgradea
         revert NonDelegatable();
     }
 
-    /// @notice Get the balance of a specific LP for a given account
-    /// @param _holder The address of the account to get the balance for
-    /// @param _lp The address of the LP to get the balance for
-    /// @return LPData memory The balance of the LP for the given account
+    /// @notice Get the balance and scaling factor of a specific LP for a given account
+    /// @param _holder The address of the account to get the data for
+    /// @param _lp The address of the LP to get the data for
+    /// @return LPData memory The balance and scaling factor of the LP for the given account
     function balanceOfLP(address _holder, address _lp) external view returns (LPData memory) {
         return _accountToLPData[_holder][_lp];
     }
 
     /// @notice Deposit LP tokens and mint ButteredBread with corresponding LP scaling factor
     function _deposit(address _account, address _lp, uint256 _amount) internal {
-        IERC20(_lp).transferFrom(_account, address(this), _amount);
-        _accountToLPData[_account][_lp].balance += _amount;
-
-        uint256 currentScalingFactor = scalingFactors[_lp];
-        _accountToLPData[_account][_lp].scalingFactor = currentScalingFactor;
-
-        _mint(_account, _amount * currentScalingFactor / FIXED_POINT_PERCENT);
         _syncDelegation(_account);
 
+        /// @dev ensure proper accounting in case of admin error in `modifyScalingFactor` where not all holders are updated
+        _syncVotingWeight(_account, _lp);
+        _accountToLPData[_account][_lp].balance += _amount;
+        
+        _mint(_account, _amount * scalingFactors[_lp] / FIXED_POINT_PERCENT);
+        
+        bool success = IERC20(_lp).transferFrom(_account, address(this), _amount);
```

Comment:
Why is transfer moved after mint? Funds should arrive to this contract, then tokens minted to the depositor.

Reply:
to follow the checks-effects-interactions pattern.
it's in the [commit message](https://github.com/BreadchainCoop/breadchain/pull/120/commits/c210616e12eeab396cb9ed15d3ab17cab125ab8c) :)

Reply:
to follow the checks-effects-interactions pattern, which reduces potential vulnerabilities like reentrancy.
and anyway it all happens in the same transactions so it either all happens or it all doesn't happen.. the order doesn't change the outcome.
actually to make the review process easier, i submitted issues for each fix and linked them to each commit that fixes them. this one fixes #116
welcome to look at the rest :)
#114 #115 #117 #118 #119

Reply:
Here is a nice blog posts on this [pattern](https://samvishwas.medium.com/solidity-programming-the-checks-effects-interactions-pattern-bac0e5a5f659#:~:text=The%20first%20step%20in%20the,proceed%20with%20the%20requested%20action.)

Reply:
@RonTuretzky wrote a article on this too few years ago 
https://medium.com/better-programming/solidity-smart-contract-security-preventing-reentrancy-attacks-fc729339a3ff

Reply:
I get it is to follow the CEI pattern, but that usually applies to funds flowing out to the caller. in this case, funds are flowing into the contract, only then will coins be minted to the caller. so in this context, it seems minting first before receiving payment is more of a vulnerability, right?

Reply:
If I'm understanding correctly, what we're afraid of is a re-entrancy attack from the LP. 

If the re-entrancy entrypoint is in the `transferFrom` , it means the attack uses `transferFrom` to call `deposit` again

so this is the attack in the currently deployed implementation

```mermaid 
sequenceDiagram 
Attacker ->> BB : deposit
loop  
BB ->> CorruptedLP : transferFrom
CorruptedLP ->> BB : deposit
end
BB ->> BB : mint
```

I believe the proposed change would make it so that this is possible instead 

```mermaid 
sequenceDiagram 
Attacker ->> BB : deposit
loop  
BB ->> CorruptedLP : transferFrom
BB ->> BB : mint
CorruptedLP ->> BB : deposit
end
```

Which seems worst to me

Reply:
In the second view, the mint would occur before the transferFrom-deposit sequence. But the point is understood.

Notice in the _withdraw function, the balance is updated before being sent to the caller, whereas in the deposit function the balance is updated after the funds are received to the contract. It is the same pattern but applied in reverse depending on the context.

receive funds from caller => mint coins to caller
burn coins from caller => release funds to caller

Reply:
updated the diagram to reflect the mint before transferFrom

Reply:
Might be worth it to add a re-entrancy guard tho

Reply:
great catch! sorry guys, didn't think this through.. good lesson for me that best practices are not always best :))
i will revert this change and add a reentrancy guard 🙏

Reply:
moved back in https://github.com/BreadchainCoop/breadchain/pull/120/commits/4fc9568be5c3ac64a0e28e082bf3ebf407e27761 and added reentrancy guard in https://github.com/BreadchainCoop/breadchain/pull/120/commits/73bc7fc447979812b2856913bea23733ef0a755b
didn't add on the `withdraw` function yet, on one hand can't hurt, but it will add some gas cost (to reduce the gas cost we can implement it using [`transient` which is supposed to reduce the gas cost significantly](https://docs.soliditylang.org/en/v0.8.28/contracts.html#transient-storage), but i'm guessing OZ will update their implementation to it soon)

Reply:
I guess we should add it to the withdraw flow as well, as there could be a re-entrancy as well over there if we're already covering it

Reply:
added, so resolving this :)